### PR TITLE
Villager specific positions stored in entity

### PR DESCRIPTION
**Pull request by sfence**
This is an experimental change, how villagers can go away from building signs. So, more villagers can sleep in one building with more beds. This should be easily extendable. So, for example, food_pos can represent tavern in village and etc. So it can be helpful to create a working village.

screenshot_20210530_213859
screenshot_20210530_213911

### CODE REVIEW

**Comment by theFox6** [comment]
Originally I planned a detection for beds, chests, etc. and the Building sign was supposed to manage upgrading of buildings. If villagers manage that themselves that might be a little bit more dynamic though.

**Comment by sfence** [comment]
Yes. I go around building signs because this solution allows more villagers to live in one house with more beds.
But, Building signs can be still used by villagers to load positions (Not implemented now).

**Comment by scrain777** [comment]
This is an interesting concept, but seems like it would be difficult to manage manually for even a moderately sized village.

Food_pos: Maybe villagers could exchange the positions of various common amenities when they are working near each other. That could include tools, food, locations of other resources. Or, maybe this should be managed at the village level (e.g. town hall)?

Bed_pos: I am thinking that if we want to support multiple beds per building (very sensible), it would make sense to store the door position in the bed's metadata rather than storing them both in the villager. That could reduce bugs related to mismatches where the door is in one house and the bed in another. (Or maybe it just substitutes errors in villager data with errors in bed data?)  For a scalable interface, we might want to add a feature to the town hall where it shows all available villagers and all available beds and has a drag-and-drop for assigning beds.

**Comment by sfence** [comment]
> This is an interesting concept, but seems like it would be difficult to manage manually for even a moderately sized village.
> 
> Food_pos: Maybe villagers could exchange the positions of various common amenities when they are working near each other. That could include tools, food, locations of other resources. Or, maybe this should be managed at the village level (e.g. town hall)?
> 
> Bed_pos: I am thinking that if we want to support multiple beds per building (very sensible), it would make sense to store the door position in the bed's metadata rather than storing them both in the villager. That could reduce bugs related to mismatches where the door is in one house and the bed in another. (Or maybe it just substitutes errors in villager data with errors in bed data?) For a scalable interface, we might want to add a feature to the town hall where it shows all available villagers and all available beds and has a drag-and-drop for assigning beds.

Some management will have to be added in the future.
But it will be nice, to keep here the possibility for players, to instruct villagers.
My idea is to add the job "city manager" in the future.

**Comment by theFox6** [comment]
Due to the matching problem I decided to make building markers and I planned to have those store all the information about a building, like beds, door/s, storages, etc. and I planned to hand off the creation of new buildings to the town hall (with optional mayor).

**Comment by sfence** [comment]
My suggestion:

- Keep position data of villager in villager entity.
- Add possibility to a villager to be owned by village (not by player)
- Change loading of villager data (for villager from the village, position data will be loaded as a link to the table in village data, otherwise, from entity static data)
- add some village data manager (maybe like a folder in the world directory, with one file for each village - prevent loading/saving one big file)
- add city/village manager/mayor to take care of village data (like add new houses etc.)
- make something like mayor table to make players able to check village data or do it via speaking with the village mayor.

This will make as able to keep possible for players to command villagers fully manually and the mayor to take care about his village. And for debugging, it will be still easy to check villager position data. This function can be reserved for the villager's owner and admins.

**Comment by sfence** [comment]
So. Now, villagers store pos data in itself. There is preparation for using the village data manager in the future. It also includes compatibility with building markers.

Synchronization is done by sharing table pointers. So necessary changes in marker update bed/door pos functions has been done.

**Comment by sfence** [comment]
screenshot_20211121_170035

Now, data change form support village name change as well.

Building markers are backward compatible. So, villagers from previous versions should be able to see the marker door and bed position still.

I expect village management to be done later in different branches/commits.

So please check it, give me feedback, and if it will look well, i believe, it can be merged into master.

File: working_villagers/forms.lua [hunk 1]
```diff
@@ -225,101 +234,259 @@ forms.register_page("working_villages:job_change",{
...
-    -- home position
-		local home_pos = {x = 0, y = 0, z = 0}
+local function floor_pos(pos)
+	pos.x = math.floor(pos.x)
+	pos.y = math.floor(pos.y)
+	pos.z = math.floor(pos.z)
+	return pos
+end
+local function load_pos(pos, villager, nodes)
+	if (pos=="near") then
```
Comment (theFox6) [inline working_villagers/forms.lua:244]: very good idea

**Review by theFox6 — APPROVED** [review]

**Pull request merged**